### DISCUSSION TITLE
Group by and Filter on many csv.gz files leads to excessive memory consumption

### DISCUSSION THREAD

**Discussion by DonIvanCorleone** [duckdb/duckdb#9215] [Q&A] [answered]
Hello there,

I am uncertain if DuckDB can be used for this use case or not: I will try to explain as good as possible:

 I have a couple (~50K) compressed `csv.gz` files laying around in a directory. Now I would like to do the following:
`SELECT * FROM (SELECT *, row_number() OVER (PARTITION by 'filename', 'column_a' ORDER BY 'column_b' ) as rn FROM read_csv_auto('directory/*.csv.gz', filename=true, ignore_errors=true)) where rn <= 35` So basically I want to have the first `n` elements of each group per file. 
In plain Python I can achieve such a thing within memory limits but exessive wait time. Thus I believed duckDB for the win: Unfortunately I am a poor guy and do not own so much memory as duckDB would like to "swallow" for this task - and my free disk space is not sufficient either (~150GB)

Is there a way to enforce duckDB to exceute the SQL  "more file based" as it is right now? 

Hope anyone has an idea? Or maybe can explain why this is so costly (in terms of memory) in duckDB. 

Many thanks in advance

Daniel

**Comment by l1t1** [comment]
try set threads=1, if it still oom, try use pypy instead of python

**Reply by Mause** [reply]
DuckDB doesn't support pypy, and it would not help here anyway

**Reply by DonIvanCorleone** [reply]
Hi @l1t1 ,

thanks a lot for your suggestion. Unfortunately, this did not help too much. It basically extended the time until the memory was full and the spillage to disk began and in the very end: OOM kicked in. 

PyPy is not an option for me either. 

Any additional ideas?

**Reply by DonIvanCorleone** [reply, marked as answer]
Just for clarification and if somebody stumbles across some similar problem:

I "solved" my problem by making use of Hive partitioned *.parquet files (https://duckdb.org/docs/data/partitioning/hive_partitioning). First I streamed from my *.csv.gz files into *.parquet files (https://duckdb.org/docs/data/partitioning/partitioned_writes) then I adopted my queries to respect the partitioning. Finally I was able to reduce the amount of required memory to something I could offer. 

Thats at least the way I could circumvent my problem and potentially its helpful for someone else aswell.

**Comment by l1t1** [comment]
I mean use pypy to complete the process, not combine with duckdb, it often faster than cpython and use less memory
